### PR TITLE
Merging additional features to master branch

### PR DESCRIPTION
**Pull request by saurabhprasun20**
Features to add:

1. Dynamic loading of obstacles.
2. Plotting graph for tree and apartment obstacles. 
3. Uploading logs and other files as zip folder to the cloud.
4. K8s image load from .env file.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,18 +6,38 @@ RUN sudo apt-get update -y &&\
...
 WORKDIR /src/aerialist/
 RUN pip3 install -r requirements.txt
+RUN pip2 install roslibpy
+RUN pip3 install roslibpy
 COPY . .
 RUN chmod +x ./aerialist/__main__.py
 COPY ./template.env ./.env
 RUN mkdir -p /io/ ./results/logs/ tmp/
 
+RUN cd /home &&\
```
Comment (saurabhprasun20) [inline Dockerfile:35]: Added the code to download all the required gazebo models from this repo. I need to add the gazebo model of tree and apartment to load it into the simulator.

File: aerialist/px4/drone_test.py [hunk 2]
```diff
@@ -148,6 +151,7 @@ def plot(
                 obstacles=None
                 if test.simulation is None
                 else test.simulation.obstacles,
+                upload_dir=test.agent.path,
                 filename=filename,
             )
```
Comment (saurabhprasun20) [inline aerialist/px4/drone_test.py:156]: Added the upload_dir as another param to the method to upload the plot into the modified test.agent.path folder. Since the Trajectory class doesn't have Dronetest as an attribute, I passed this as another parameter.

File: aerialist/px4/k8s_agent.py [hunk 3]
```diff
@@ -96,7 +99,7 @@ def run(self):
                 logger.debug("files downloaded")
 
             for test_log in os.listdir(local_folder):
-                if test_log.endswith(".ulg") and (
+                if ".ulg" in test_log and (
```
Comment (saurabhprasun20) [inline aerialist/px4/k8s_agent.py:102]: Since we are zipping the file, we longer have file name ending with .ulg

File: aerialist/px4/obstacle.py [hunk 4]
```diff
@@ -124,15 +146,29 @@ def to_dict(self):
                 "x": self.position.x,
                 "y": self.position.y,
                 "z": self.position.z,
-                "r": self.position.r,
+                "angle": self.position.r,
```
Comment (skhatiri) [inline aerialist/px4/obstacle.py:149]: update this, and related code to "r" to be consistant

File: aerialist/px4/obstacle.py [hunk 5]
```diff
@@ -124,15 +146,29 @@ def to_dict(self):
...
         }
 
     @classmethod
     def distance_to_many(cls, obstacles: List[Obstacle], line: LineString):
         boxes = [o.geometry for o in obstacles]
         dist = min([sum([b.distance(Point(*p)) for b in boxes]) for p in line.coords])
-        return dist
+        dist_list = [sum([b.distance(Point(*p)) for b in boxes]) for p in line.coords]
+        dist_avg = sum(dist_list) / len(dist_list)
+        return dist, dist_list
```
Comment (skhatiri) [inline aerialist/px4/obstacle.py:160]: adding new return values to this function could break current code, let's check together

File: aerialist/px4/obstacle.py [hunk 6]
```diff
@@ -21,26 +22,37 @@ class Position(NamedTuple):
...
         self,
         size: Size,
         position: Position,
+        # angle: float = 0,
         shape: str = BOX,
     ) -> None:
         super().__init__()
-
-        if shape == self.BOX:
+        if shape == "BOX" or shape == "APARTMENT":
```
Comment (skhatiri) [inline aerialist/px4/obstacle.py:36]: use constants like above for comparison

File: aerialist/px4/simulator.py [hunk 7]
```diff
@@ -57,6 +53,7 @@ def __init__(self, config: SimulationConfig) -> None:
                 f"DONT_RUN=1 make -C {self.PX4_DIR} px4_sitl_default gazebo; "
             )
             sim_command += f". {self.PX4_DIR}Tools/setup_gazebo.bash {self.PX4_DIR} {self.PX4_DIR}build/px4_sitl_default; "
+            # sim_command += f". {self.PX4_DIR}Tools/simulation/gazebo/setup_gazebo.bash {self.PX4_DIR} {self.PX4_DIR}build/px4_sitl_default; "
```
Comment (skhatiri) [inline aerialist/px4/simulator.py:56]: remove line

File: aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
```
Comment (skhatiri) [inline aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py:1]: I would move all the folder to resources/simulation/

File: aerialist/ros_node_script/spawn_script/spawn_pattern_box.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            <geometry>
+              <box>
+                <size>{0} {1} {2}</size>
+              </box>
+            </geometry>
+          </collision>
+          <visual name='visual'>
+            <geometry>
+              <box>
+                <size>{1} {2} {3}</size>
```
Comment (skhatiri) [inline aerialist/ros_node_script/spawn_script/spawn_pattern_box.py:57]: {0} {1} {2} seems to be right here

File: aerialist/ros_node_script/spawn_script/spawn_pattern_box.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    # Initialize the ROS node
+    rospy.init_node(box_name)
+    # rospy.spin()
+
+    # Wait for the Gazebo spawn service
+    rospy.wait_for_service("/gazebo/spawn_sdf_model")
+    spawn_sdf_model = rospy.ServiceProxy("/gazebo/spawn_sdf_model", SpawnModel)
+
+    # Define the SDF model for a box with the specified size
+    box_sdf = """
```
Comment (skhatiri) [inline aerialist/ros_node_script/spawn_script/spawn_pattern_box.py:29]: why not put this texts (also for other obstacles) in a .sdf file instead of hardcoded here?
let's discuss it

File: aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            z,
+            r,
+        ]
+    )
+
+
+def launch_node():
+    # Iterate over the array and launch nodes
+    rospy.init_node("dynamic_obstacle_launcher")
+    list1 = rospy.get_param("/dynamic_obstacle_launcher/obstacle_string").split(",end,")
```
Comment (skhatiri) [inline aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py:55]: we could use a better delimitter. or better format. like: [,,,,],[,,,],[,,,,],[,,,,]. let's discuss this

File: aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+import subprocess
```
Comment (skhatiri) [inline aerialist/ros_node_script/spawn_script/spawn_dynamic_obstacles.py:2]: why don't you call this code (and all related code) directyly in simulator.py? we could have a seperate "gazebo / ros" class that wraps all the details of putting obstacles in the simulator.
is there any issues with that? let's discuss it
it is now being called by the launch file, which makes it hard to follow the code flow.

File: aerialist/resources/simulation/collision_prevention_bkp.world [hunk 13]
```diff
@@ -0,0 +1,83 @@
+<sdf version='1.6'>
```
Comment (skhatiri) [inline aerialist/resources/simulation/collision_prevention_bkp.world:1]: remove the file if unsued

File: aerialist/px4/trajectory.py [hunk 14]
```diff
@@ -245,7 +269,7 @@ def plot_multiple(
             fig.savefig(plot_file)
             plt.close(fig)
             if cls.WEBDAV_DIR is not None:
-                file_helper.upload(f"{cls.DIR}{filename}.png", cls.WEBDAV_DIR)
```
Comment (skhatiri) [inline aerialist/px4/trajectory.py:248]: cls.WEBDAV_DIR is set and used in so many palces. better to keep using the same variable for the moment

File: aerialist/px4/trajectory.py [hunk 15]
```diff
@@ -114,9 +117,30 @@ def plot_multiple(
         xy_plt.set_aspect("equal", "datalim")
 
         if obstacles is not None:
+            labeled = False
+            labeled_tree = False
+            labeled_apartment = False
             for obst in obstacles:
-                obst_patch = obst.plt_patch()
-                obst_patch.set_label("obstacle")
+                if obst.shape == "BOX":
```
Comment (skhatiri) [inline aerialist/px4/trajectory.py:124]: use the constants defined for the shape instead of strings for comparison

File: aerialist/px4/simulator.py [hunk 16]
```diff
@@ -65,18 +62,31 @@ def __init__(self, config: SimulationConfig) -> None:
...
-                if len(self.config.obstacles) > 2:
-                    sim_command += f"obst3:=true obst3_x:={self.config.obstacles[2].position.y} obst3_y:={self.config.obstacles[2].position.x} obst3_z:={self.config.obstacles[2].position.z} obst3_l:={self.config.obstacles[2].size.w} obst3_w:={self.config.obstacles[2].size.l} obst3_h:={self.config.obstacles[2].size.h} obst3_yaw:={math.radians(-self.config.obstacles[2].position.r)} "
-                if len(self.config.obstacles) > 3:
-                    sim_command += f"obst4:=true obst4_x:={self.config.obstacles[3].position.y} obst4_y:={self.config.obstacles[3].position.x} obst4_z:={self.config.obstacles[3].position.z} obst4_l:={self.config.obstacles[3].size.w} obst4_w:={self.config.obstacles[3].size.l} obst4_h:={self.config.obstacles[3].size.h} obst4_yaw:={math.radians(-self.config.obstacles[3].position.r)} "
-
-        logger.debug("executing:" + sim_command)
+            sim_command += f"exec roslaunch {self.AVOIDANCE_LAUNCH} gui:={str((not self.config.headless) and self.GAZEBO_GUI_AVOIDANCE).lower()} rviz:={str(True and not self.config.headless).lower()} world_file_name:={self.config.world_file_name[0]} "
```
Comment (skhatiri) [inline aerialist/px4/simulator.py:65]: let's discuss and see if we can find a less complicated way of invoking obstacle placement code. I put some suggestions in the spawn_script folder

File: aerialist/px4/simulator.py [hunk 17]
```diff
@@ -120,8 +130,8 @@ def start(self) -> bool:
                 if self.config.simulator == SimulationConfig.ROS and output.endswith(
                     "INFO  [tone_alarm] home set"
                 ):
-                    logger.info("Avoidance is ready (waiting 5 seconds to wrap up)")
-                    time.sleep(5)
+                    logger.info("Avoidance is ready (waiting 10 seconds to wrap up)")
+                    time.sleep(10)
```
Comment (skhatiri) [inline aerialist/px4/simulator.py:134]: we can define it as a variable on top. this depends on how many obstacles are being added, right?

File: aerialist/px4/local_agent.py [hunk 18]
```diff
@@ -25,7 +26,14 @@ def run(self):
                 self.drone.run_scheduled()
             log = self.simulator.get_log()
             if self.config.agent is not None and self.config.agent.path is not None:
-                file_helper.upload(log, self.config.agent.path)
+                upload_string = config("UPLOAD_LIST", default="")
```
Comment (skhatiri) [inline aerialist/px4/local_agent.py:29]: let's keep uploading the unzipped log to the cloud. 
let's stract a helper function for the whole thing and put in the filehelper.

File: aerialist/px4/k8s_agent.py (hunk 3 above)
Comment (skhatiri) [reply aerialist/px4/k8s_agent.py:102]: let's keep the .ulg file as is. don't zip it

File: aerialist/px4/file_helper.py [hunk 19]
```diff
@@ -142,7 +143,7 @@ def upload_dir(src_dir: str, dest_dir: str) -> str:
                 return upload_dir(src_dir, dest_dir)
             else:
                 raise (e)
-    return dest_dir
+        return dest_dir
```
Comment (skhatiri) [inline aerialist/px4/file_helper.py:146]: if it doe's break your code, let's keep this as before
.

File: aerialist/px4/drone_test.py (hunk 2 above)
Comment (skhatiri) [reply aerialist/px4/drone_test.py:156]: can't we use WEBDAV_DIR for it?

File: aerialist/px4/drone_test.py [hunk 20]
```diff
@@ -87,24 +91,23 @@ def cmd_params(self):
...
-                        params += f"{p} "
-                if len(self.simulation.obstacles) >= 3:
-                    params += "--obstacle3 "
-                    for p in self.simulation.obstacles[2].to_params():
-                        params += f"{p} "
-                if len(self.simulation.obstacles) >= 4:
-                    params += "--obstacle4 "
-                    for p in self.simulation.obstacles[3].to_params():
-                        params += f"{p} "
+            if (
```
Comment (skhatiri) [inline aerialist/px4/drone_test.py:94]: i think you have missed some new updates in the merge here

File: aerialist/px4/drone_test.py [hunk 21]
```diff
@@ -219,6 +224,7 @@ def __init__(
         self.headless = headless
         self.obstacles: List[Obstacle] = obstacles
         self.home_position = home_position
+        self.world_file_name = None
```
Comment (skhatiri) [inline aerialist/px4/drone_test.py:227]: what is this compared to the "world" above? i think you can use the same variable, it is not much used yet as far as i remember

File: samples/tests/mission1-k8s.yaml [hunk 22]
```diff
@@ -1,29 +1,84 @@
 drone:
   port: ros #{sitl, ros, cf}
-  params_file: samples/flights/mission1-params.csv #csv file with the same structure as above
```
Comment (skhatiri) [inline samples/tests/mission1-k8s.yaml:3]: revert the changes to keep this sample simple. create another sample to demonstrate complidated obstacle types

File: samples/tests/mission1-local.yaml [hunk 23]
```diff
@@ -0,0 +1,55 @@
+drone:
```
Comment (skhatiri) [inline samples/tests/mission1-local.yaml:1]: i think i renamed this sample to merge local and docker samples. you can remove it if not needed in your code

File: Dockerfile [hunk 24]
```diff
@@ -6,18 +6,38 @@ RUN sudo apt-get update -y &&\
...
+RUN echo "export ROS_PACKAGE_PATH=/src/aerialist/aerialist/inter_images:\$ROS_PACKAGE_PATH" >> /root/.bashrc
+
+RUN sudo apt install ros-melodic-stereo-image-proc ros-melodic-image-view -y
+# RUN sudo apt-get install python-rospy
 #&&\
 # sudo -H pip3 install --upgrade pip
 
 #Setting up the current tool 
 COPY ./requirements.txt /src/aerialist/requirements.txt
+WORKDIR /src/catkin_ws/src/avoidance/
```
Comment (skhatiri) [inline Dockerfile:18]: let's see if we still need this parts after possible changes to the obstacle spawn code

File: Dockerfile [hunk 25]
```diff
@@ -6,18 +6,38 @@ RUN sudo apt-get update -y &&\
...
+WORKDIR /src/catkin_ws/src/avoidance/spawn_obstacles/src
+RUN mkdir -p nodes
+WORKDIR /src/catkin_ws/src/avoidance/spawn_obstacles/src/nodes
+COPY aerialist/ros_node_script/spawn_script/ .
+RUN catkin build -w /src/catkin_ws &&\
+    echo "source /src/catkin_ws/devel/setup.bash" >> ~/.bashrc  
 WORKDIR /src/aerialist/
 RUN pip3 install -r requirements.txt
+RUN pip2 install roslibpy
+RUN pip3 install roslibpy
```
Comment (skhatiri) [inline Dockerfile:29]: why not add to requirements.txt?

File: PX4-Dockerfile [hunk 26]
```diff
@@ -1,5 +1,7 @@
 FROM px4io/px4-dev-ros-melodic:2021-09-08
 
+RUN mkdir /data
```
Comment (skhatiri) [inline PX4-Dockerfile:3]: let's skip this and move to the other dockerfile if not neccesary here

File: requirements.txt [hunk 27]
```diff
@@ -11,9 +11,11 @@ webdavclient3==3.14.6
 ruptures==1.1.7
 shapely==1.8.4
 validators==0.20.0
+opencv-python==4.8.0.74
 pyyaml==6.0
 munch==2.5.0
 rospkg==1.4.0
 defusedxml==0.7.1
+tslearn== 0.5.2
```
Comment (skhatiri) [inline requirements.txt:19]: duplicate

File: template.env [hunk 28]
```diff
@@ -8,15 +8,23 @@ ROS_HOME=/root/.ros/
 LOGS_COPY_DIR=results/
 # A folder to save the experiment results
 RESULTS_DIR=results/
+THRESHOLD_FILE=threshold.csv
```
Comment (skhatiri) [inline template.env:11]: are these needed here?

File: template.env [hunk 29]
```diff
@@ -31,19 +39,25 @@ ALLIGN_ORIGIN=True
...
 
-
 # DOCKER
 DOCKER_IMG=skhatiri/aerialist
 DOCKER_TIMEOUT=1000
 # SIMULATION_TIMEOUT=970
 
 # k8s
-# ROS_KUBE_TEMPLATE=aerialist/resources/k8s/k8s-job-avoidance.yaml
+ROS_KUBE_TEMPLATE=aerialist/resources/k8s/k8s-job-avoidance-pattern.yaml
```
Comment (skhatiri) [inline template.env:48]: keep your specific configuration defaults in a seperate file or comment (#) and describe them here

File: template.env [hunk 30]
```diff
@@ -31,19 +39,25 @@ ALLIGN_ORIGIN=True
...
 # KUBE_TEMPLATE=aerialist/resources/k8s/k8s-job.yaml
 
 # webdav server
 # WEBDAV_DL_FLD=/tmp/
 # WEBDAV_UP_FLD=tmp/
+WEBDAV_UP_FLD=webdav://test/
```
Comment (skhatiri) [inline template.env:54]: #

**Review by skhatiri — COMMENTED** [review]
Thanks for the effort.
I added some comments. you can update the code, and we can discuss some of them in detail to make the code ready for merge.

File: aerialist/px4/obstacle.py (hunk 5 above)
Comment (saurabhprasun20) [reply aerialist/px4/obstacle.py:160]: I will create a new method.

File: aerialist/px4/simulator.py [hunk 31]
```diff
@@ -65,18 +62,31 @@ def __init__(self, config: SimulationConfig) -> None:
...
+            tree_count = 0
+            apartment_count = 0
+            box_count = 0
+            obstacle_string = ""
+            if self.config.obstacles is not None and len(self.config.obstacles) > 0:
+                for obstacle in self.config.obstacles:
+                    if obstacle.shape == "BOX":
+                        box_count += 1
+                        box_name = "box_" + str(box_count)
+                        obstacle_string += f"{obstacle.position.y},{obstacle.position.x},{obstacle.position.z},{math.radians(-obstacle.position.r)},box,{box_name},{obstacle.size.w},{obstacle.size.l},{obstacle.size.h},end,"
```
Comment (skhatiri) [inline aerialist/px4/simulator.py:75]: obstacle.shape

File: aerialist/px4/simulator.py (hunk 17 above)
Comment (saurabhprasun20) [reply aerialist/px4/simulator.py:134]: I will make thema a variable

File: PX4-Dockerfile (hunk 26 above)
Comment (saurabhprasun20) [reply PX4-Dockerfile:3]: cehck if we can remove this.

File: template.env (hunk 28 above)
Comment (saurabhprasun20) [reply template.env:11]: removed
